feature_69_lulc_use_case (#81)
NCAR/i-wrf: docs/Users_Guide/lulcncarhpc.rst

:orphan:

.. _lulcncarhpc:

***************************************************************
Running I-WRF on NCAR HPCs with Land Use/Land Cover Change Data
***************************************************************

Overview
========

> Update section with an overview of the following instructions.

Reviewer: I see a few notes to update the sections with content on this page. Should this be replaced with the actual instructions?

It is recommended that you follow the instructions in each section in the order presented to avoid encountering issues during the process. This guide assumes that I-WRF LULC simulations have already been completed and focuses specifically on the METplus verification workflow.

Prepare to Use NCAR HPCs
========================

To get started with running METplus verification on NCAR HPC systems, you will need access to one of the supported platforms (Derecho, Casper, or other NCAR computing resources). If you do not already have access, you will need to request an account through the NCAR Computing and Information Systems Laboratory (CISL).

NCAR HPC access is typically provided to researchers affiliated with NCAR, university collaborators, or those with approved allocations. If you need to request access:

* Visit the NCAR CISL user registration page to create an account
* Provide information about your research project and institutional affiliation
* Wait for account approval, which typically takes 1-2 business days
* Once approved, you can log in to the HPC systems using SSH

> Most NCAR HPC systems use a batch job scheduling system for computational work. While this exercise can be run interactively for demonstration purposes, production runs should be submitted through the job scheduler. Refer to the NCAR HPC documentation for specific guidance on your target system.

Reviewer 2: ```suggestion
Most NCAR HPC systems use a batch job scheduling system for computational work. While this exercise can be run interactively for demonstration purposes, production runs should be submitted through the job scheduler. Refer to the `NCAR HPC documentation <https://arc.ucar.edu/docs>`_ for specific guidance on your target system.
```

Once you have access to an NCAR HPC system, you can log in using SSH from your local machine or through NCAR's web-based interfaces where available.

Preparing the Environment
=========================

With your NCAR HPC account active and you logged in to the system, you can now set up the environment and create the necessary directories to run METplus verification. You will only need to perform these steps once per system, as the directory structure and configuration will remain available for future use.

The following sections instruct you to issue numerous Linux commands in your shell. The commands in each section can be copied and pasted into your shell terminal.

> 

Reviewer 2: ```suggestion
Launch Bash Shell
-----------------

The following command are formatted to run in the Linux Bash shell, which you may or may not be running by default. Copy and paste the following commands to print your current shell and then launch a new Bash shell, if needed::

   echo $0
   bash

```

Load Required Modules
---------------------

NCAR HPC systems use environment modules to manage software. Load the Apptainer module which provides the containerization software needed to run METplus::

   module load apptainer

> 

Reviewer 2: Surprisingly, running `module load` in tcsh prior to switching to `bash` seems to work. `apptainer` is still in my path after switching shells.
```suggestion
Confirm Bash Shell
------------------

The following commands are formatted to run in the Linux Bash shell, which you may or may not be running by default. Copy and paste the following command to print your default shell::

   echo $SHELL

If needed, switch to the :code:`bash` shell by running::

   bash

Note that you can modify your default login shell through the `Systems Account Manager (SAM) <https://sam.ucar.edu/>`_ but changes can take up to several hours to take effect.
```

Define Environment Variables
----------------------------

We will be using environment variables throughout this exercise to ensure consistent file paths and resource names. Copy and paste the definitions below into your shell before proceeding::

    IWRF_WORK_DIR=${SCRATCH}/iwrf_work
    LOCAL_OUTPUT_DIR=${IWRF_WORK_DIR}/metplus_out
    export APPTAINER_TMPDIR=${TMPDIR}

> Any time you open a new shell session on the HPC system, you will need to reload the apptainer module and redefine these variables before executing the commands that follow.

Reviewer 2: ```suggestion
Any time you open a new shell session on the HPC system, you will need to reload the apptainer module, switch shells, if needed, and redefine these variables before executing the commands that follow.
```

This creates a local copy of all I-WRF configuration files, including the METplus settings needed for the LULC verification workflow.

Pull Apptainer Objects
======================

As mentioned above, the METplus software is provided as a containerized image that will run using Apptainer on your NCAR HPC system. Apptainer (formerly Singularity) is the preferred containerization technology on HPC systems, as it provides secure container execution without requiring root privileges. Unlike cloud environments that use Docker directly, NCAR HPC systems use Apptainer to run containerized applications.

The METplus image contains all the necessary software and dependencies to perform verification of I-WRF LULC simulation output. You can "pull" (download) the METplus image from the container registry to your HPC system's storage.

Get the METplus and Data Container Images
-----------------------------------------

You must pull the METplus software container and the input data containers that contain the observational and WRF simulation data for the LULC use case::

   apptainer pull ${IWRF_WORK_DIR}/iwrf-metplus.sif docker://ncar/iwrf-metplus:latest

>    apptainer pull ${IWRF_WORK_DIR}/data-lulc-input-obs.sif docker://ncar/iwrf-data:lulc-input-obs-d03.apptainer

Reviewer 2: As discussed over Slack, something has changed which requires us to pull data images through `oras://registry-1.docker.io` rather than `docker://`. This is that change.
```suggestion
   apptainer pull ${IWRF_WORK_DIR}/data-lulc-input-obs.sif oras://registry-1.docker.io/ncar/iwrf-data:lulc-input-obs-d03.apptainer
```

   apptainer pull ${IWRF_WORK_DIR}/data-lulc-input-wrf.sif docker://ncar/iwrf-data:lulc-input-wrf-d03.apptainer

> These commands download three container images: the METplus software, the observational data, and the WRF simulation data. The process may take several minutes depending on your network connection.

Reviewer 2: ```suggestion
These commands download three container images: the METplus software, the observational data, and the WRF simulation data. Processing the METplus software can take up to 15 minutes and the data containers can take several minutes each.
```

This configuration provides the container with access to:

* Observational data from the ``data-lulc-input-obs.sif`` container image at ``/data/input/obs``
* WRF simulation data from the ``data-lulc-input-wrf.sif`` container image at ``/data/input/wrf``
* METplus configuration files from the I-WRF repository at ``/config``
* Visualization script files for generating plots at ``/plot_scripts``
* Output directory for writing verification results at ``/data/output``
* Temporary directory for Apptainer operations

Execute METplus Verification
----------------------------

The LULC use case includes verification for two meteorological variables: accumulated precipitation and radar reflectivity. Each verification is run separately using its own METplus configuration file.

Run the accumulated precipitation verification::

   apptainer exec ${IWRF_WORK_DIR}/iwrf-metplus.sif /metplus/METplus/ush/run_metplus.py /config/GridStat_apcp_lulc.conf

> This process compares simulated precipitation accumulation against observational data and generates statistical metrics. Progress information is displayed while the verification is performed.

Reviewer 2: ```suggestion
This process compares simulated precipitation accumulation against observational data and generates statistical metrics. While this may take up to 20 minutes to run, progress information is displayed while the verification is performed.
```

Run the reflectivity verification::

   apptainer exec ${IWRF_WORK_DIR}/iwrf-metplus.sif /metplus/METplus/ush/run_metplus.py /config/GridStat_refc_lulc.conf

> This process evaluates the model's ability to simulate radar reflectivity patterns compared to observed radar data.

Reviewer 2: ```suggestion
This process evaluates the model's ability to simulate radar reflectivity patterns compared to observed radar data. This step may take up to 10 minutes to run.
```

Both verification processes use GridStat, which computes grid-to-grid verification statistics. The tools generate comprehensive statistical output including bias, correlation, and skill scores that quantify the model's performance.

Verify Output Generation
------------------------

After both METplus runs complete successfully, you can verify that the output files were created properly.

Check that the GridStat verification output was generated::

>    ls ${LOCAL_OUTPUT_DIR}/grid_stat/* -1

Reviewer 2: Putting the `-1` option at the end looks weird.
```suggestion
   ls -1 ${LOCAL_OUTPUT_DIR}/grid_stat/*
```
